### PR TITLE
Replace ctime with ctime_r

### PR DESCRIPTION
**Pull request by JelteF**
ctime is a banned function

### CODE REVIEW

File: src/bin/pg_autoctl/state.c [hunk 1]
```diff
@@ -492,29 +494,32 @@ NodeStateFromString(const char *str)
...
+ *
+ * This string is stored in buffer and for ease of use of the function this
+ * buffer is also returned. The buffer should (at least) be MAXCTIMESIZE large.
  */
 const char *
-epoch_to_string(uint64_t seconds)
+epoch_to_string(uint64_t seconds, char *buffer)
 {
-	if (seconds > 0)
+	if (seconds <= 0) {
```
Comment (DimCitus) [inline src/bin/pg_autoctl/state.c]: Style issue: brackets on their own line.

File: src/bin/pg_autoctl/state.c [hunk 2]
```diff
@@ -492,29 +494,32 @@ NodeStateFromString(const char *str)
...
-			return NULL;
-		}
-		str[strlen(str) - 1] = '\0';
-		return str;
+		log_error("Failed to convert epoch %" PRIu64 " to string: %m",
+				  seconds);
+		return NULL;
 	}
-	return "0";
+	buffer[strlen(buffer) - 1] = '\0';
```
Comment (DimCitus) [inline src/bin/pg_autoctl/state.c]: Can the buffer be an empty string or something else that would trip us here? It looks suspicious the way we address the memory here.

File: src/bin/pg_autoctl/state.c [hunk 3]
```diff
@@ -492,29 +494,32 @@ NodeStateFromString(const char *str)
...
+	if (seconds <= 0) {
+		strlcpy(buffer, "0", MAXCTIMESIZE);
+		return buffer;
+	}
+	/*
+	 * ctime_r returns a string that ends with \n, which we don't want.
+	 * Replace it with a null string terminator, even if that means we lose
+	 * a precious byte of memory here..
+	 */
+	ctime_r((time_t *) &seconds, buffer);
```
Comment (DimCitus) [inline src/bin/pg_autoctl/state.c]: When discarding the return value of a function call, we often write `(void) funcall(...)` to make it obvious to the readers: we are discarding it on purpose.

File: src/bin/pg_autoctl/state.c (hunk 2 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/state.c]: I now see that we used to do it that way already, so I guess it's fine. Still looks suspicious... but it does not belong to the current PR to solve it I guess.

File: src/bin/pg_autoctl/state.c (hunk 3 above)
Comment (JelteF) [reply src/bin/pg_autoctl/state.c]: I'm actually using the result now, since it indicates if it errors.

File: src/bin/pg_autoctl/state.c (hunk 2 above)
Comment (JelteF) [reply src/bin/pg_autoctl/state.c]: Added some sanity checks there.

File: src/bin/pg_autoctl/state.c [hunk 4]
```diff
@@ -492,29 +494,35 @@ NodeStateFromString(const char *str)
...
+	}
+	result = ctime_r((time_t *) &seconds, buffer);
+
+	if (result == NULL)
 	{
+		log_error("Failed to convert epoch %" PRIu64 " to string: %m",
+				  seconds);
+		return NULL;
+	}
+	if (strlen(result) != 0 && result[strlen(result) - 1] == '\n') {
```
Comment (DimCitus) [inline src/bin/pg_autoctl/state.c]: Style issue: brackets on their own line.

**Pull request merged**